sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLoggingPolicy.java

```diff
@@ -194,32 +209,45 @@ private HttpLoggingPolicy(HttpLogOptions httpLogOptions, boolean prettyPrintJson
 
         addHeadersToLogMessage(response.getHeaders(), responseLogMessage, numericLogLevel);
 
-        Mono<String> responseLoggingMono = Mono.defer(() -> Mono.just(responseLogMessage.toString()));
-
-        if (httpLogDetailLevel.shouldLogBody()) {
-            final String contentTypeHeader = response.getHeaderValue("Content-Type");
+        if (!httpLogDetailLevel.shouldLogBody()) {
+            responseLogMessage.append("<-- END HTTP");
+            return logAndReturn(logger, responseLogMessage, response);
+        }
 
-            if (shouldBodyBeLogged(contentTypeHeader, getContentLength(logger, response.getHeaders()))) {
-                final HttpResponse bufferedResponse = response.buffer();
-                responseLoggingMono = bufferedResponse.getBodyAsString().flatMap(body -> {
+        String contentTypeHeader = response.getHeaderValue("Content-Type");
+        long contentLength = getContentLength(logger, response.getHeaders());
+
+        if (shouldBodyBeLogged(contentTypeHeader, contentLength)) {
+            HttpResponse bufferedResponse = response.buffer();
+            ByteArrayOutputStream outputStream = new ByteArrayOutputStream((int) contentLength);
+            return bufferedResponse.getBody()
+                .doOnNext(byteBuffer -> {
+                    for (int i = byteBuffer.position(); i < byteBuffer.limit(); i++) {
```

Comment:
Same here. Should we have support for truncating body to log only the first few bytes?

Reply:
Response in your other question.